src/app/utils/getRPCAccountBalances.ts

```diff
@@ -0,0 +1,47 @@
+import { Runtime, RuntimeSdkBalance } from '../../oasis-nexus/api'
+import * as oasis from '@oasisprotocol/client'
+import * as oasisRT from '@oasisprotocol/client-rt'
+import { Network } from '../../types/network'
+import { getTokensForScope, paraTimesConfig } from '../../config'
+import { fromBaseUnits } from './helpers'
+
+const grpcUrls: { [key in Network]: string } = {
+  [Network.mainnet]: 'https://grpc.oasis.dev',
+  [Network.testnet]: 'https://testnet.grpc.oasis.dev',
+}
+
+export async function getRPCAccountBalances(
+  oasisAddress: string,
+  scope: {
+    network: Network
+    layer: Runtime
+  },
+): Promise<RuntimeSdkBalance[]> {
+  const paratimeConfig = paraTimesConfig[scope.layer][scope.network]
+  if (!paratimeConfig.runtimeId) throw new Error('Paratime is not configured')
+
+  const nic = new oasis.client.NodeInternal(grpcUrls[scope.network])
+  const accountsWrapper = new oasisRT.accounts.Wrapper(oasis.misc.fromHex(paratimeConfig.runtimeId))
+
+  const rawBalances = (
+    await accountsWrapper
+      .queryBalances()
+      .setArgs({ address: oasis.staking.addressFromBech32(oasisAddress) })
+      .query(nic)
+  ).balances
+
+  return [...rawBalances.entries()]
+    .map(([denomination, amount]) => {
+      return {
+        balance: oasis.quantity.toBigInt(amount).toString(),
+        token_symbol: oasis.misc.toStringUTF8(denomination) || getTokensForScope(scope)[0].ticker,
+        token_decimals: paraTimesConfig[scope.layer].decimals, // TODO: differentiate by token
+      }
+    })
+    .map(token => {
+      return {
+        ...token,
+        balance: fromBaseUnits(token.balance, token.token_decimals),
+      }
+    })
```

Comment:
This is iterating twice over an array, for no good reason. This should happen in the previous `map`.

Reply:
I prefer this, and it mirrors https://github.com/oasisprotocol/explorer/blob/26079807ff9a4e91cc42d106d75a0c824533caba/src/oasis-nexus/api.ts#L361-L366. 

(I also think https://github.com/oasisprotocol/explorer-frontend/blob/26079807ff9a4e91cc42d106d75a0c824533caba/src/oasis-nexus/api.ts#L709-L752 should be split into multiple iterations over an array)

Reply:
I know it is not a large array, but for me it makes no sense, to iterate over the same array multiple times.